### ISSUE TITLE
Associations across different databases

### ISSUE THREAD

**Issue by Zigur** [balderdashy/waterline#779]
I have two models, ```Subject``` and ```PersonalDetails``` that are living on two different PostgreSQL databases (i.e. connections). ```Subject``` has a one-to-one association to  ```PersonalDetails``` through its attribute ```personalInfo```

```
var Subject= {

    attributes: {

        code: {
            type:'string',
            required: true,
            unique: true
            //alphanumeric: true
        },

        sex: {
            type: 'string',
            required: true,
            enum: ['M', 'F', 'N.D.', 'UNKNOWN', 'UNDIFFERENTIATED']
        },

        // one-way association to PersonalDetails model
        personalInfo: {
            model: 'personalDetails',
            columnName: 'personal_info'
        }
    }
};
```
If I run a query on Subjects populating personalInfo: 
```
sails> Subject.findOne(1).populate('personalInfo').exec(function(err, res) {
... if (err) console.log(err);
... else console.log(res);
... });
```

I get this error.

```
sails> Error (E_UNKNOWN) :: Encountered an unexpected error
: Unable to determine primary key for collection `personal_details` because an error was encountered acquiring the collection definition:
[TypeError: Cannot read property 'definition' of undefined]
    at _getPK (/home/massi/Projects/xtens-app/node_modules/sails-postgresql/lib/adapter.js:926:13)
    at StrategyPlanner.__FIND__.Cursor.$getPK (/home/massi/Projects/xtens-app/node_modules/sails-postgresql/lib/adapter.js:504:20)
    at StrategyPlanner.determineStrategy (/home/massi/Projects/xtens-app/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/planner.js:90:22)
    at /home/massi/Projects/xtens-app/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/planner.js:33:25
    at Array.forEach (native)
    at StrategyPlanner.plan (/home/massi/Projects/xtens-app/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/planner.js:30:31)
    at Cursor.run (/home/massi/Projects/xtens-app/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/cursor.js:39:30)
    at runJoins (/home/massi/Projects/xtens-app/node_modules/sails-postgresql/node_modules/waterline-cursor/index.js:51:10)
    at __FIND__ (/home/massi/Projects/xtens-app/node_modules/sails-postgresql/lib/adapter.js:478:9)
    at after (/home/massi/Projects/xtens-app/node_modules/sails-postgresql/lib/adapter.js:965:7)

Details:  Error: Unable to determine primary key for collection `personal_details` because an error was encountered acquiring the collection definition:
[TypeError: Cannot read property 'definition' of undefined]
```

This seems a duplicate of this one (https://github.com/balderdashy/sails-mysql/issues/150), but the issue seemed to have been fixed on waterline 0.10.13.
I am using sails-postgres 0.10.9 (PostgreSQL 9.4) and waterline 0.10.14.

Any suggestion? Should I manually do the association for all the Subjects? Thank you so much.

**Comment by particlebanana** [comment]
Does the personal details model have any kind of special primary key or is it just using the autoPk?

**Comment by Zigur** [comment]
Just using the autoPk.

**Comment by appdevdesigns** [comment]
I'm having a similar problem with the latest npm versions:
sails @ 0.10.5
waterline: @0.10.16
sails-mysql: @0.10.11

Similar setup:  2 Models using connections from 2 different databases (sails-mysql).

FCFMinistry    has many   FCFActivities:

FCFMinistry:
```javascript
module.exports = {

    tableName:"tblMinistries",
    autoCreatedAt:false,
    autoUpdatedAt:false,
    autoPK:false,
    migrate:'safe',  // don't update the tables!

    connection:"fcf",

    attributes: {

        IDMinistry : {
            type : "integer",
            size : 10,
            primaryKey : true
        }, 

////  rest removed for brevity

        // @hasMany Activities (FCFActivity)
        activities:{
            collection:'FCFActivity',
            via: 'team'
        }

    }
};
```

FCFActivity:
```javascript
module.exports = {

    tableName:"fcf_activity",

    connection:"mysql",

migrate:'alter',  

    attributes: {

        date_start : { type: 'date' },

        date_end : { type: 'date' },

////  other fields removed for brevity

        // @hasOne Team (FCFMinistry)
        team:{
            model:'FCFMinistry'
        }
          

    }
};
```

Then on the sails console I get this:
```
sails> FCFMinistry.find({IDMinistry:4}).populate('activities').exec(console.log);
undefined
sails> Error (E_UNKNOWN) :: Encountered an unexpected error
: Unable to determine primary key for collection `fcf_activity` because an error was encountered acquiring the collection definition:
[TypeError: Cannot read property 'definition' of undefined]
    at _getPK (/Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/lib/adapter.js:1164:13)
    at StrategyPlanner.__JOIN__.Cursor.$getPK (/Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/lib/adapter.js:544:20)
    at StrategyPlanner.determineStrategy (/Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/node_modules/waterline-cursor/cursor/planner.js:90:22)
    at /Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/node_modules/waterline-cursor/cursor/planner.js:33:25
    at Array.forEach (native)
    at StrategyPlanner.plan (/Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/node_modules/waterline-cursor/cursor/planner.js:30:31)
    at Cursor.run (/Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/node_modules/waterline-cursor/cursor/cursor.js:39:30)
    at runJoins (/Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/node_modules/waterline-cursor/index.js:51:10)
    at __JOIN__ (/Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/lib/adapter.js:518:9)
    at afterwards (/Users/codingMonkey/Sites/develop/sails-plugins-v10/node_modules/sails-mysql/lib/connections/spawn.js:84:5)

Details:  Error: Unable to determine primary key for collection `fcf_activity` because an error was encountered acquiring the collection definition:
[TypeError: Cannot read property 'definition' of undefined]
```
 
FCFMinistry has a nonstandard primary key definition.  

FCFActivity is using the sails default (id).

But the problem is on the FCFActivity reference.  


If I go in and change FCFActivity to be the same adaptor, then all works fine.

```
sails> FCFMinistry.find({IDMinistry:4}).populate('activities').exec(console.log);
undefined
sails> null [ { activities: 
     [ { date_start: null,
         date_end: null,
         team: 4,
         id: 1,
         createdAt: null,
         updatedAt: null },
       { date_start: null,
         date_end: null,
         team: 4,
         id: 2,
         createdAt: null,
         updatedAt: null } ],
    IDMinistry: 4,
    codeMinistryStatus: 'A',
    IDProject: '...',
    NameMinistryNat: '...',
    NameMinistryEng: '...',
    DateMinistryStarted: null,
    DateMinistryEnded: null,
    IDPersonResponsible: null,
    WorkAddressThai: null,
    WorkAddressEng: null,
    MinistryDisplayName: '...' } ]
```

Things seem to be crashing in ``` /sails-mysql/lib/adapter.js:1169:13 ``` 
on this line:
```javascript 
collectionDefinition = connections[connectionIdentity].collections[collectionIdentity].definition; 
```

if I insert 
```javascript 
console.log(_.keys(connections[connectionIdentity].collections));
 ```

I can see that collections is populated with all the table names of the adaptor for FCFMinistry.  But none of the ones for the adaptor for FCFActivity.

Hope this helps ...

**Comment by dfulgham** [comment]
Im seeing this issue in waterline 0.10.18 as well.  

I have two connections defined Conn1 and Conn2

When I call Conn1.findOne(1).populate('assocField').exec(console.log) for example.  I see that Waterline asks Conn1's adapter to find schema and primaryKey for Conn1's model, but then also asks Conn1's adapter to find schema and primaryKey for Conn2's model, and this is where the error occurs.

I placed a console.log at the following line in the sails-mysql "version": "0.10.11" in  /lib/adapter.js file

(line 506)
    join: function (connectionName, collectionName, options, cb, connection) {
	console.log("MYSQL:",collectionName, options);
...
(line 542) 
    $getPK: function (collectionName) {
	          console.log("MYSQL $getPK",collectionName);
...

And the following is logged, indicating that the initial connection is being used to attempt to connect to the second database incorrectly.

Conn1.find(1).populate('assocField').exec(console.log)
MYSQL: tenantMaster { where: { id: 1 },
  joins: 
   [ { parent: 'tenantMaster',
       parentKey: 'capsule',
       child: 'party',
       childKey: 'id',
       select: [Object],
       alias: 'capsule',
       removeParentKey: true,
       model: true,
       collection: false,
       criteria: [Object] } ] }

 MYSQL $getPK conn1
_findPK firstName
_findPK lastName
_findPK assocField
_findPK id
MYSQL $getPK conn2
Error (E_UNKNOWN) :: Encountered an unexpected error
: Unable to determine primary key for collection `conn2` because an error was encountered acquiring the collection definition:
[TypeError: Cannot read property 'definition' of undefined]
    at _getPK (/Users/user/Documents/GIT Projects/test/node_modules/sails-mysql/lib/adapter.js:1166:13)
    at StrategyPlanner.__JOIN__.Cursor.$getPK (/Users/user/Documents/GIT Projects/test/node_modules/sails-mysql/lib/adapter.js:545:20)
...
etc.

**Comment by devjones** [comment]
+1. Experiencing a similar issue.

Campaign Model is in my primary database.  Connection established with sails-postgresql adapter
autoPK: true,

Reference to geonames:
```
    geonames: {
      collection: 'geoname',
      dominant: true
    },
```

Geoname Model is in a *separate* database. Connection established with sails-postgresql adapter.
```autoPK: false```

Primary key setup:
```
    geonameId: {
      columnName: 'geonameid',
      type: 'integer',
      primaryKey: true
    },
```

When I attempt the following command, I get an error:
```Campaign.find().populate('geoname').exec(console.log);```

Full error stack trace:
```
Error (E_UNKNOWN) :: Encountered an unexpected error
: Unable to determine primary key for collection `geoname` because an error was encountered acquiring the collection definition:
[TypeError: Cannot read property 'definition' of undefined]
    at _getPK (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/lib/adapter.js:999:13)
    at StrategyPlanner.__FIND__.Cursor.$getPK (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/lib/adapter.js:543:20)
    at StrategyPlanner.determineStrategy (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/planner.js:91:22)
    at /Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/planner.js:34:25
    at Array.forEach (native)
    at StrategyPlanner.plan (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/planner.js:30:31)
    at Cursor.run (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/waterline-cursor/cursor/cursor.js:39:30)
    at runJoins (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/waterline-cursor/index.js:51:10)
    at __FIND__ (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/lib/adapter.js:515:9)
    at after (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/lib/adapter.js:1131:7)
    at /Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/lib/adapter.js:1025:7
    at /Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/pg/lib/pool.js:77:9
    at dispense (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/pg/node_modules/generic-pool/lib/generic-pool.js:250:16)
    at Object.me.acquire (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/pg/node_modules/generic-pool/lib/generic-pool.js:319:5)
    at Object.pool.connect (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/pg/lib/pool.js:71:12)
    at PG.connect (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/node_modules/pg/lib/index.js:49:8)
    at spawnConnection (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/lib/adapter.js:1024:8)
    at Object.module.exports.adapter.join (/Users/devx/code/projects/livechat-server/node_modules/sails-postgresql/lib/adapter.js:511:7)
    at module.exports.join (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/adapter/dql.js:55:13)
    at _runOperation (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:401:29)
    at run (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/operations.js:69:8)
    at bound.module.exports.find (/usr/local/lib/node_modules/sails/node_modules/waterline/lib/waterline/query/finders/basic.js:243:16)

Details:  Error: Unable to determine primary key for collection `geoname` because an error was encountered acquiring the collection definition:
[TypeError: Cannot read property 'definition' of undefined]
```

**Comment by devinivy** [comment]
Sounds like there's a bug here!  Marking as such.  @particlebanana any thoughts about this?  Thoughts on if this a problem with the integrator or query portions of waterline core?

**Comment by dmarcelino** [comment]
@devjones, would you mind trying the changes from PR #789 (branch: [bradshja:master](https://github.com/bradshja/waterline.git)), it has been reported to fix https://github.com/balderdashy/waterline/issues/833#issuecomment-74275620 - Join failing across adapters.

**Comment by dmarcelino** [comment]
Should be fixed by #925 / #984. Please re-open if this re-surfaces.
